enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,456 @@
+desirable/possible in all situations.
+
+Using `NetworkPolicy` a user may control the traffic that is allowed into
+the cluster. By creating a deny-all policy the router will still remain
+operative for internal traffic, while rejecting all external traffic. An
+example follows:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: block-router
+  namespace: openshift-ingress
+spec:
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 10.42.0.0/16
+  podSelector:
+    matchLabels:
+      ingresscontroller.operator.openshift.io/deployment-ingresscontroller: default
+  policyTypes:
+  - Ingress
+```
+Assuming the pod network is configured to use CIDR `10.42.0.0/16`, the above
+policy will block all traffic not coming from that network.
+
+Any of these methods is equivalent to using firewalld to block traffic, but
+using the CNI instead of a host level setting. This provides greater
+flexibility, as it is self contained within MicroShift APIs.
+
+If a user needs to block traffic to a port this needs a `NetworkPolicy`
+resource like the one shown above. This is NOT automatically done by
+MicroShift, user action is required.
+
+#### Exposing on hostnames
+Relying on [LoadBalancer controller enhancement](loadbalancer-service-support.md)
+means the services are updated by this component. However, ovnk only turns IP
+addresses into iptables rules, it does not take Hostname, which is another
+valid field for the service status.
+
+Using hostnames to expose the router may imply the use of DNS. Since ovnk only
+accepts IP addresses, the controller must resolve those names to their
+corresponding IPs, then configure them in the service. In order to do this, it
+must periodically check the IPs for the names to keep them updated.
+
+#### Allowing specific IP addresses
+Allowing specific IP addresses is achieved through the use of `NetworkPolicy`
+resources, as seen in previous examples. These policies include the allowed IP
+addresses and ports in the spec, meaning they reject everything else that is
+not included explicitly.
+
+Due to MicroShift's networking setup, there is one more change required for
+the `NetworkPolicy` to filter out based on the source IP.
+
+As it can be seen in the iptables rules, traffic towards the router ports is
+subject to DNAT to the router service IP and port. All services use
+`externalTrafficPolicy: Cluster`, which will SNAT traffic's source IP to match
+that of the node. This gets further NAT, as described here. By the time the
+`NetworkPolicy` is evaluated the source IP has been replaced with
+`100.64.0.0/16`, which will not produce matches against the configured IP in
+the policy, and traffic will get rejected.
+
+To overcome this, all LoadBalancer services in MicroShift that are subject to
+get NetworkPolicy resources to control external traffic must set
+`externalTrafficPolicy: Local`. This setting will not SNAT packets and the
+source IP is preserved until NetworkPolicy checks run.
```

Comment:
How will we ensure that all the right services are configured this way? Does it only apply to the router service, or will we need to make additional changes?

Reply:
This only applies to LoadBalancer services that need to perform ingress traffic control using NetworkPolicy. I think documenting it should be enough?

Reply:
Will this be different from host a similar LoadBalancer service on OpenShift would need to be configured?

Reply:
I would say it shouldnt. Configuring a LB service means there is an external LB that will later on do the balancing towards different nodes. There is no need to perform further balancing within the cluster. OCP does have a LB before the nodes, and microshift does not need one, so the situation is equivalent.